lifecycle.go

```diff
@@ -40,14 +45,104 @@ type Lifecycle interface {
 type Hook struct {
 	OnStart func(context.Context) error
 	OnStop  func(context.Context) error
+
+	onStartName string
+	onStopName  string
 }
 
-type lifecycleWrapper struct{ *lifecycle.Lifecycle }
+// StartHook returns a new Hook with start as its [Hook.OnStart] function,
+// wrapping its signature as needed. For example, given the following function:
+//
+//	func myhook() {
+//	  fmt.Println("hook called")
+//	}
+//
+// then calling:
+//
+//	lifecycle.Append(StartHook(myfunc))
+//
+// is functionally equivalent to calling:
+//
+//	lifecycle.Append(fx.Hook{
+//	  OnStart: func(context.Context) error {
+//	    myfunc()
+//	    return nil
+//	  },
+//	})
+//
+// The same is true for all functions that satisfy the HookFunc constraint.
+// Note that any context.Context parameter or error return will be propagated
+// as expected. If propagation is not intended, users should instead provide a
+// closure that discards the undesired value(s), or construct a Hook directly.
+func StartHook[T HookFunc](start T) Hook {
+	onstart, startname := lifecycle.Wrap(start)
+
+	return Hook{
+		OnStart:     onstart,
+		onStartName: startname,
+	}
+}
+
+// StopHook returns a new Hook with stop as its [Hook.OnStop] function,
+// wrapping its signature as needed. For example, given the following function:
+//
+//	func myhook() {
+//	  fmt.Println("hook called")
+//	}
+//
+// then calling:
+//
+//	lifecycle.Append(StopHook(myfunc))
+//
+// is functionally equivalent to calling:
+//
+//	lifecycle.Append(fx.Hook{
+//	  OnStop: func(context.Context) error {
+//	    myfunc()
+//	    return nil
+//	  },
+//	})
+//
+// The same is true for all functions that satisfy the HookFunc constraint.
+// Note that any context.Context parameter or error return will be propagated
+// as expected. If propagation is not intended, users should instead provide a
+// closure that discards the undesired value(s), or construct a Hook directly.
+func StopHook[T HookFunc](stop T) Hook {
+	onstop, stopname := lifecycle.Wrap(stop)
+
+	return Hook{
+		OnStop:     onstop,
+		onStopName: stopname,
+	}
+}
+
+// StartStopHook returns a new Hook with start as its [Hook.OnStart] function
+// and stop as its [Hook.OnStop] function, independently wrapping the signature
+// of each as needed.
+func StartStopHook[T1 HookFunc, T2 HookFunc](start T1, stop T2) Hook {
```

Comment:
I'm not sure if this was considered, but what if we automatically wrap OnStart and OnStop hooks in fx.Hook passed to `lifecycle.Append` and just update the type definition for fx.Hook to take in the new types of function that we're allowing? I think that will eliminate the need for these helper functions, right? I'm just thinking if we ever add another type of hook to lifecycle, we would have to add a bunch of these helper functions again. Is there performance or backward compatibility issue that I'm not aware of with this approach?

Reply:
I had considered that first. Unfortunately, we can't update `fx.Hook` to do this without parameterizing it to use the `HookFunc` type constraint, which would (1) qualify as a breaking change and (2) make direct `fx.Hook` much less ergonomic:

```go
type Hook[T1 HookFunc, T2 HookFunc] struct {
  OnStart T1
  OnStop  T2
}
```

which, at a minimum, would require both (1) explicit parameterization for struct literals and (2) both type parameters to be specified even if only using one field.

Reply:
I see. Thank you for explaining this to me!